tests/as_ref.rs

```diff
@@ -452,6 +681,28 @@ mod multi_field {
             assert!(ptr::eq(rf, item.first.as_ref()));
         }
 
+        #[derive(AsRef)]
+        struct Types {
+            #[as_ref(str)]
+            first: String,
```

Comment:
@tyranron 
I don't think this syntax can work on stable.

Autoref-based specialization gets most of the way there, but can't handle generics.

For instance, without something to differentiate these 2 cases, there's no way of generating correct impl bounds for both

```rust
// where T: TypeEquals<Other = T>
#[derive(AsRef)]
#[as_ref(T)]
struct Foo<T>(T);

// where T: AsRef<i32>
#[derive(AsRef)]
#[as_ref(i32)]
struct Bar<T>(T);
```

As an alternative, we could allow multiple attributes

```rust
#[derive(AsRef)]
#[as_ref]
#[as_ref(str)]
struct Foo(String);
```

Or use a special keyword for the field type

```rust
#[derive(AsRef)]
#[as_ref(str, direct)]
struct Foo(String);
```

Either one would remove the need for specialization.

Reply:
@MegaBluejay 

> Autoref-based specialization gets most of the way there, but can't handle generics.

But do we need to?

> ```rust
> #[derive(AsRef)]
> #[as_ref(T)]
> struct Foo<T>(T);
> ```

This seems to be trivial, as we cannot alias a type parameter. So we can just check whether it contains in the type generics and treat this situation as `#[as_ref]`.

Reply:
One case I don't think we could handle this way is when the field type is itself generic and uses a type parameter

```rust
#[derive(AsRef)]
#[as_ref(i32)]
struct Foo<T>(T);

#[derive(AsRef)]
#[as_ref(Foo<T>, T)]
struct Bar<T>(Foo<T>);
```

`&Bar<T> -> &T` should be forwarded, even though there's a type parameter involved.

I'll look into this some more and figure out if there's a good way of working around this

Reply:
@MegaBluejay can you show the expansion with autoref-based specialization for this case with the required types machinery?

Reply:
@tyranron [something like this](https://gist.github.com/MegaBluejay/945b19de936a76783b52bb29d4229082)

Reply:
@MegaBluejay

First, I don't understand all the squats around `Cell`. Seems totally redundant and misleading, so I've removed it. It's not required for type-level reasoning (coherence) at all.

Second, your example seems to work just fine with autoref-based specialization, even when a type alias is inclined: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b54d0c62dea54208fe910fea5abfdb9f).

Third, we may omit using autoref-based specialization in the cases where the specified type is named exactly the same as the one of the field:
```rust
#[derive(AsRef)]
#[as_ref(Vec<T>)]
struct Stack<T>(Vec<T>);

// expands to just:
impl<T> AsRef<Vec<T>> for Stack<T> {
    fn as_ref(&self) -> &Vec<T> {
        &self.0
    }
}
```
We need only involve autoref-based specialization machinery when some "outer" types are used.

Given all of these, it seems that we're covering all the possible cases here. If not, prove me wrong with an example that doesn't compile on playground, please.

Reply:
@MegaBluejay ah yes, of course, we cannot conditionally generate that bound:
```rust
where
    Foo<T>: AsRef<T>,
```
We either can generate it for both impls, which breaks the first impl, or doesn't generate it at all, which breaks the second impl.

Let me think over it a little bit... it seems that the problem is not fundamental, and we can overcome it by bounding with `Conv<_>: ExtractRef` somehow, which will automatically imply the necessary conditions.

Reply:
@MegaBluejay 

> `Cell` is required for the `AsMut` version, because `Conv` needs to contain a mutable reference.

Yes, variance gives there some hard times for mutable references. Let's keep separate machinery traits for `AsRef` and `AsMut` versions, so we don't bother `AsRef` version with runtime impact of a `Cell`. Regarding `AsMut`, for now, use `Cell` in machinery, but give some more time poking around to make it working without `Cell`. There should be a way, imho.

> As for bounding with `Conv`, I don't think it's possible, since code using autoref-specialization isn't actually valid for all its cases at once, it just compiles differently in different contexts, which can't be expressed in types

True. We cannot express "OR" bounds, only "AND", so bringing all the bounds, we disjunct both cases, rather than conjunct them.

But we can use some trickery here, inspecting the type names. The whole algorithm would look like something like this:
1. We inspect the provided type to match by name the field type, if it does, we just use trivial `#[as_ref]` logic without specialization. It will easily handle all the cases of `impl<T> AsRef<T> for T`
    ```rust
    #[derive(AsRef)]
    #[as_ref(Vec<T>)]
    struct Stack<T>(Vec<T>);

    // expands to just:
    impl<T> AsRef<Vec<T>> for Stack<T> {
        fn as_ref(&self) -> &Vec<T> {
            &self.0
        }
    }
    ```
    except the cases where type is renamed
    ```rust
    type V<T> = Vec<T>;

    #[derive(AsRef)]
    #[as_ref(V<T>)]
    struct Stack<T>(Vec<T>);
   ```
2. If the provided type doesn't match the field type, we use the specialization wrapper.
3. If the provided type contains one of type parameters of the declared type (we can recursively inspect it), we add the `FieldType: AsRef<ProvidedType<T>>` bound, which enables the second specialization impl to work in generic contexts. Even more, such as we choose here the second specialization impl only, we may not use specialization totally here, falling into the assumption, that if the types named differently - they are different types.

This way, the only situations that don't work would be when we convert directly to the field type, which involves generics, and is named differently in field and attribute:
```rust
type V<T> = Vec<T>;

#[derive(AsRef)]
#[as_ref(V<T>)]
struct Stack<T>(Vec<T>);
```

I would say that this is a very rare situation. Why would one name differently __the  same type in the same place__? Moreover, we should state explicitly in the docs, that this concrete edge case won't work as expected.

Note, that we still have specialization working okay when generics are not involved, allowing to use renamed types.

-----------------------------------------

And another direction that is worth a shot, is another kind of specialization: runtime same-type specialization, which [inspects a type to be the same in runtime and transmutes](https://github.com/graphql-rust/juniper/blob/b375146466bbd19a3c4954e20d272e46842c04be/juniper/src/value/mod.rs#L145-L163), but `TypeId` and `any` machinery is [limited to `'static` lifetime only](https://doc.rust-lang.org/stable/std/any/struct.TypeId.html#method.of), unfortunately, which would likely limit kinds of types we can place the macro onto very-very unpractically much.

Reply:
I wrote out some more impls manually following this algorithm

We'd need to omit specialization in a wider number of cases, both when the provided type contains type parameters, and when the field type does.

For instance, this case

```rust
#[derive(AsRef)]
#[as_ref(i32)]
struct Foo<T>(T);
```

Specialization can't work here either, so we must assume `i32 != T` and use a forwarded impl.

Other than that everything seems to work.